cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+                                                                                                 scheduleName));
+    if (subscriber != null) {
+      return subscriber.scheduleTaskState(program, programType, scheduleName);
+    } else {
+      return ScheduleState.NOT_FOUND;
+    }
   }
 
-  private String getScheduleId(Id.Program program, SchedulableProgramType programType, String scheduleName) {
-    return String.format("%s:%s", getProgramScheduleId(program, programType), scheduleName);
+  private synchronized Store getStore() {
+    if (store == null) {
+      store = storeFactory.create();
+    }
+    return store;
   }
 
-  private String getProgramScheduleId(Id.Program program, SchedulableProgramType programType) {
-    return String.format("%s:%s:%s:%s", program.getNamespaceId(), program.getApplicationId(),
-                         programType.name(), program.getId());
+  /**
+   * One instance of this class contains a list of {@link StreamSizeSchedule}s, which are all interested
+   * in the same stream. This instance subscribes to the size notification of the stream, and polls the
+   * stream for its size whenever the schedules it references need the information.
+   * The {@link StreamSizeScheduler} communicates with this class, which in turn communicates to the schedules
+   * it contains to perform operations on the schedules - suspend, resume, etc.
+   */
+  private final class StreamSubscriber extends AbstractScheduledService
+    implements NotificationHandler<StreamSizeNotification> {
+    // Key is the schedule ID
+    private final ConcurrentMap<String, StreamSizeScheduleTask> scheduleTasks;
+    private final Id.Stream streamId;
+    private final AtomicInteger activeTasks;
+
+    private Cancellable notificationSubscription;
+    private StreamSizeNotification lastNotification;
+    private StreamSize lastPollingInfo;
+    private Long delta;
+
+    private StreamSubscriber(Id.Stream streamId) {
+      this.streamId = streamId;
+      this.scheduleTasks = Maps.newConcurrentMap();
+      this.activeTasks = new AtomicInteger(0);
+      this.delta = null;
+    }
+
+    @Override
+    protected void startUp() throws Exception {
+      notificationSubscription = notificationService.subscribe(getFeed(), this, sendPollingInfoExecutor);
+    }
+
+    @Override
+    protected void shutDown() throws Exception {
+      if (notificationSubscription != null) {
+        notificationSubscription.cancel();
+      }
+    }
+
+    @Override
+    protected void runOneIteration() throws Exception {
+      if (activeTasks.get() == 0) {
+        return;
+      }
+
+      try {
+        StreamSize streamSize = pollOnce();
+        sendPollingInfoToActiveTasks(new StreamSizeNotification(streamSize.getTimestamp(), streamSize.getSize()));
+      } catch (Throwable t) {
+        LOG.error("Could not poll stream {}", streamId.getName(), t);
```

Comment:
The message is not correct as you don't know whether the throwable is from the polling (pollOnce()) or from sending it to the active tasks.